extractdilkey.c

```diff
@@ -0,0 +1,386 @@
+                    break;
+                }
+            }
+            else
+            {
+                printf("**** ERROR: Unrecognized raw private key : %s\n", sInFile);
+                sRc = 1;
+                break;
+            }
+        }
+        else
+        {
+            // Attempt to convert encoded key
+            unsigned int sWireType = 0;
+            if (sPubOut)
+            {
+                // Get the raw public key
+                sRc = mlca_wire2key(sRawKey, sRawBytes, &sWireType,
+                                    sKey, sKeyBytes,
+                                    (const unsigned char *) CR_OID_SPECIAL_PRV2PUB,
+                                    CR_OID_SPECIAL_PRV2PUB_BYTES);
+                if (0 >= sRc)
+                {
+                    printf("**** ERROR: Unable to convert private key : %d\n",sRc);
+                    sRc = 1;
+                    break;
+                }
+                else
+                {
+                    sRawBytes = sRc;
+                    sRc = 0;
+                }
+                if (sVerbose) printf("extractdilkey: Found public key\n");
+
+                // Encode it
+                sRc = mlca_key2wire(sKey,    sKeyBytes,
+                                    sRawKey, sRawBytes, 0,
+                                    NULL, 0,
+                                    NULL, 0);
+                if (sRc < 0)
+                {
+                    printf("**** ERROR: Failure during public key conversion : %d\n", sRc);
+                    break;
+                }
+                sKeyBytes = sRc;
+            }
+            else
+            {
+                sRc = mlca_wire2key(sRawKey, sRawBytes, &sWireType,
+                                    sKey, sKeyBytes,
+                                    NULL, ~0);
+
+                // Raw private key size for dilithium r2 8/7
+                if (0 >= sRc || 5136 != sRc)
+                {
+                    printf("**** ERROR: Unable to convert private key : %d\n",sRc);
+                    sRc = 1;
+                    break;
+                }
+                else
+                {
+                    sRawBytes = sRc;
+                    sRc = 0;
+                }
+            }
+        }
+
+        if (NULL != sOutFile)
+        {
+            printf("Writing private key to : %s\n", sOutFile);
+            if (sRawOut)
+            {
+                sRc = writeFile(sRawKey, sRawBytes, sOutFile);
+            }
+            else
+            {
+                sRc = writeFile(sKey, sKeyBytes, sOutFile);
+            }
+        }
+        else
+        {
+            printf("Valid Dilithium private keyfile detected\n");
+        }
+    }
+    } while(0);
+
+    exit(sRc);
+
+	free(sKey);
+	free(sRawKey);
+	exit(sRc);
+}
+
+
+int readFile(unsigned char *data,
+	     size_t *length,
+	     const char *filename)
+{
+	int sRc = 0;
```

Comment:
would either of these make more sense?
bool failure = false
bool success = true